docs/architecture/adr-021-protobuf-query-encoding.md

```diff
@@ -0,0 +1,262 @@
+into an ABCI query as `custom/cosmos_sdk.x.bank.v1.Query/QueryBalance`. Service handlers
+registered with `QueryRouter.RegisterService` will be routed this way.
+
+Beyond the method name, GRPC requests carry a protobuf encoded payload, which maps naturally
+to `RequestQuery.Data`, and receive a protobuf encoded response or error. Thus
+there is a quite natural mapping of GRPC-like rpc methods to the existing
+`sdk.Query` and `QueryRouter` infrastructure.
+
+This basic specification allows us to reuse protocol buffer `service` definitions
+for ABCI custom queries substantially reducing the need for manual decoding and
+encoding in query methods. 
+
+### GRPC Protocol Support
+
+In addition to providing an ABCI query pathway, we can easily provide a GRPC
+proxy server that routes requests in the GRPC protocol to ABCI query requests
+under the hood. In this way, clients could use their host languages' existing
+GRPC implementations to make direct queries against Cosmos SDK app's using
+these `service` definitions. In order for this server to work, the `QueryRouter`
+on `BaseApp` will need to expose the service handlers registered with
+`QueryRouter.RegisterService` to the proxy server implementation. Nodes could
+launch the proxy server on a separate port in the same process as the ABCI app
+with a command-line flag.
+
+### REST Queries and Swagger Generation
+
+[grpc-gateway](https://github.com/grpc-ecosystem/grpc-gateway) is a project that
+translates REST calls into GRPC calls using special annotations on service
+methods. Apps that want to expose REST queries should add `google.api.http`
+annotations to their `rpc` methods as in this example below. By convention
+HTTP method paths registered this way should correspond to the protobuf
+package paths for their respective modules, i.e. `cosmos_sdk.x.bank.v1` translates
+to `/cosmos_sdk/x/bank/v1/`.
+
+
+```proto
+// x/bank/types/types.proto
+
+service Query {
+  rpc QueryBalance(QueryBalanceParams) returns (cosmos_sdk.v1.Coin) {
+    option (google.api.http) = {
+      get: "/cosmos_sdk/x/bank/v1/balance/{address}/{denom}"
+    };
+  }
+  rpc QueryAllBalances(QueryAllBalancesParams) returns (QueryAllBalancesResponse) {
+    option (google.api.http) = {
+      get: "/cosmos_sdk/x/bank/v1/balances/{address}"
+    };
+  }
+}
+```
+
+grpc-gateway will work direcly against the GRPC proxy described above which will 
+translate requests to ABCI queries under the hood. grpc-gateway can also
+generate Swagger definitions automatically.
+
+In the current implementation of REST queries, each module needs to implement
+REST queries manually in addition to ABCI querier methods. Using the grpc-gateway
+approach, there will be no need to generate separate REST query handlers, just
+query servers as described above as grpc-gateway handles the translation of protobuf
+to REST as well as Swagger definitions.
+
+The SDK should provide CLI commands for apps to start GRPC gateway either in
```

Comment:
gRPC proxy server should probably be an optional flag yes, but enabled by default. When disabled, what happens when the operator starts a REST server?

Reply:
Without the gRPC proxy, grpc-gateway wouldn't work, unless we figure out some way for it to proxy directly to ABCI (possible but haven't looked into that).

Reply:
So the REST server won't work, correct?

Reply:
grpc-gateway won't work without a gRPC -> ABCI proxy, correct.